Improve `mike` versions

llucax opened the pull request:
This pull request introduces a significant changes to the versioning scheme for `mike`. This change aims to provide greater clarity and organization to our version management and implement correct sorting.

- `next` becomes `v0.x-dev`: This alias now aligns with our development branch.

- Pre-releases receive their own versioning: For example, `v0.7-pre` is introduced for pre-release versions.

- `latest` points to the latest stable version.

- New aliases are added for the latest pre-release (`latest-pre`) and the development branch (`latest-dev`).

They ensure that users can easily identify and access the documentation version they require, whether it's a stable release, pre-release, or the latest development build.

The versions now also provide more descriptive title, using the full tag name for tags and adding a short commit hash for development versions, so users can know exactly which version they are reading the documentation for.

Fixes #149, fixes #135, fixes #76.

llucax committed:
Make the `mkdocs` module a package

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use the new module for the page generation

We still keep the exports in the package so existing projects don't need
to necessarily update their scripts.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a script to better sort `mike` versions

This is to sort the versions displayed by mike so new release
(and develpment releases) are listed first.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Sort `mike` versions before the site is published

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add module to get version information about a repository

This handles many scenarios and queries, like:

* Is the current commit a tag or a branch?
* Is the current tag the latest tag?
* Is the current branch the latest branch?
* Is the current tag the last minor version for the major version?
* What is the next minor version for the current major branch?
* etc.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a module to deal with `mike` versions

This module provides these tools:

* Building the mike version information from the repository information
* Sorting the mike version information `version.json` file
* Comparing mike versions

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a command-line tool to get the current `mike` version

For now this tool is designed to be used in GitHub Actions workflows,
but is is possible to use it in other environments as well doing some
extra configuration.

To use this tool call it via python:

`python -m frequenz.repo.config.cli.version.mike.info`

In the future the idea will be to have a command with sub-commands, like
`repo-config version mike info`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Remove unnecessary environment variables from steps

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Update step ID that gets mike version info

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Use the new tool to get the mike version information

This means we use new aliases: `next` becomes `v0.x-dev`. Pre-releases
will get a version too (for example `v0.7-pre`). The `latest` version
now correctly points to the bigger stable version, and new aliases are
added for the latest pre-release (`latest-pre`) and development branch
(`latest-dev`).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Improve `mike` version titles

Now use the full tag name for tags, and for branches add a short SHA, so
users know the exact version of the documentation they are reading.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Move the `mike` version sorting CLI tool

The `mike` version sorting CLI tool is now in
`src/frequenz/repo/config/cli/version/mike/sort.py` instead of in
`src/frequenz/repo/config/mkdocs/sort_versions.py`.

It can be invoked with `python -m
frequenz.repo.config.cli.version.mike.sort`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update cookiecutter template to use the new `mike` tools

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
ci: Install deps before running the mike tool

We need to first install our own package to be able to use the mike CLI
tool to get the version information.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove references to `next` in the documentation

Update it to use `v0.1-dev` and `latest-dev` instead.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
I tested this in my fork and it seems to be working well:

* https://llucax.github.io/frequenz-repo-config-python/
* https://github.com/llucax/frequenz-repo-config-python/actions/runs/6533809227/job/17740272989

Marenz commented on src/frequenz/repo/config/mkdocs/sort_versions.py:
```diff
@@ -0,0 +1,159 @@
+
+    - `vX` -> `vX.9999.9999-zzzz`
+    - `vX.Y` -> `vX.Y.9999-zzzz`
+    - `vX.Y.Z` -> `vX.Y.Z`
```
I assume the `999`s and the `zz`s are there to make it sort them to the end(or front depending on order)?
Maybe a comment mentioning that part would be helpful, strictly speaking this is more "orderable_semver" no?

llucax replied:
Yeah, so the new function name should make it more evident, but I'll improve the description a bit more too.

Marenz commented on src/frequenz/repo/config/version.py:
```diff
@@ -0,0 +1,443 @@
+
+    Returns:
+        The branches dictionary, where the key is the branch name and the value is
+            the parsed branch name, sorted by branch version.
```
I am slightly confused by the `sorted` part, a `dict` isn't sorted (I thought) and otherwise and the value is a `BranchVersion` which seems to be a single value, not something that contains ordered values?

Edit: Okay, I read up a bit and it seems dict's are actually ordered in the sense that they keep their insertion order

llucax replied:
Yeah, `dict` preserve order of insertion when iterating. I'm not sure what to do here, for once it is a slightly obscure feature that not everybody knows about, but then it is just regular Python. But I guess a comment wouldn't hurt either.

llucax replied:
Mmm, looking at the code I'm not so sure about the clarification. I think adding how Python works to the documentation is too much and if I make it just a code comment then it is not adding a lot of extra clarity, especially if we have to assume the end user (only reading the docs) already should know about dicts preserving order.

llucax replied:
Oh, wait, I know see it is actually missing the sorting for the branch function (the tag has it), I guess it was lost in some rebase because I'm pretty sure I added it to both at some point.

Marenz commented on src/frequenz/repo/config/mkdocs/mike.py:
```diff
@@ -0,0 +1,250 @@
+_dev_to_semver_re = re.compile(r"^v(\d+).(\d+)-dev$")
+
+
+def _to_fake_sortable_semver(version: str) -> str:
```
Here you _did_ name the function different.

But this function and the ones below seem to be a duplication of earlier code?

llucax replied:
I'm not sure which functions you refer to, I don't see any duplicate functions, can you be more specific?

Marenz replied:
I think I got confused because I saw different names & places for the function but you actually moved & renamed it

Marenz commented:
A few comments, but LGTM

llucax committed:
Improve _to_fake_sortable_semver() description

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Really sort branches in the `version` module

To be able to properly sort branch versions, we need to handle all
special cases for versions with a `None` minor, as we can simply compare
`int`s to `None`. We consider `None` to be always bigger than any `int`
and 2 `None` to be equal.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated to clarify the `9999`s better and properly sort branches.

llucax commented:
@Marenz you enabled auto-merge but didn't approve, not sure if it was intentional 🤔
